### ISSUE TITLE
Any C++ example using MFront/MGIS ?

### ISSUE THREAD

**Issue by lilou-gautier** [thelfer/MFrontGenericInterfaceSupport#115]
Hello,

I'm completely new to MFront and MGIS but think it could be very useful inside a homemade C++ code that I can't share here.
In order to understand how everything works, I'm working with a given Norton.mfront file and try to reproduce the result of the corresponding Norton.mtest file with C++.

The .Mfront file looks like this :
```
@Parser Implicit;
@Behaviour Norton;
@Algorithm NewtonRaphson_NumericalJacobian ;

@RequireStiffnessTensor;

@MaterialProperty real A;
@MaterialProperty real m;

@StateVariable real p ;

@ComputeStress{
  sig = D*eel ;
}

@Integrator{
  real seq = sigmaeq(sig) ;
  Stensor n = Stensor(0.) ;
  if(seq > 1.e-12){
    n = 1.5*deviator(sig)/seq ;
  }
  feel += dp*n-deto ;
  fp -= dt*A*pow(seq,m) ;
} // end of @Integrator

@TangentOperator{
  Stensor4 Je ;
  getPartialJacobianInvert(Je) ;
  Dt = D*Je ;
}
```

The .mtest file looks like this :
```
@Behaviour<generic> 'lib/libBehaviour.so' 'Norton' ;
@MaterialProperty<constant> 'YoungModulus' 178600.0E6 ;
@MaterialProperty<constant> 'PoissonRatio' 0.3 ;
@MaterialProperty<constant> 'A' 8.e-67 ;
@MaterialProperty<constant> 'm' 8.2 ;
@ExternalStateVariable 'Temperature' 293.15 ;
@ImposedStress 'SXX' { 0. :  0., 30. : 40.e6};
@ImposedStress 'SXY' { 0. :  0., 30. : 40.e6};
@Times {0.,30. in 100};
```

I successfully compiled the library libBehaviour.so but I'm having difficulties using it to obtain the same result as the .mtest file.
I tried to refer to the tests files in the MGIS git but still do not manage to create a working C++ script.
It is probably very wrong but here is what I've tried yet :
```
#include <cmath>
#include <array>
#include <cstdlib>
#include <iostream>
#include <string_view>
#include "MGIS/Behaviour/Behaviour.hxx"
#include "MGIS/Behaviour/BehaviourData.hxx"
#include "MGIS/Behaviour/MaterialDataManager.hxx"
#include "MGIS/Behaviour/Integrate.hxx"
#include "TFEL/Math/stensor.hxx"


template <class T> 
void cout_vector(std::vector<T> const &input)
{
    for (int i = 0; i < input.size(); i++) {
        std::cout << input.at(i) << ' ';
    }
    //std::cout << std::endl;
}

template <typename T, size_t Size>
void cout_array(const std::array<T, Size>& arr) {
    for (const auto& element : arr) {
        std::cout << element << " ";
    }
    //std::cout << std::endl;
}

using namespace mgis;
using namespace mgis::behaviour;

int main() {

    // Load the behaviour description from the library
    const auto h = Hypothesis::TRIDIMENSIONAL;
    const auto b = load("lib/libBehaviour.so", "Norton", h);

    State state(b);

    // Set the value of the material properties
    const double E = 178600.0E6;
    const double nu = 0.3;
    const double A = 8.e-67;
    const double m = 8.2;

    setMaterialProperty(state, "YoungModulus", E);
    setMaterialProperty(state, "PoissonRatio", nu);
    setMaterialProperty(state, "A", A);
    setMaterialProperty(state, "m", m);


    // Create a BehaviourData object
    auto d = BehaviourData{b};

    // Set the initial state
    d.s0 = state;

    // Initialize time and index
    const double t_init = 0.0;  
    const double t_final = 30.0;  
    const size_type num_steps = 100;  
    const double dt = (t_final - t_init) / num_steps;  
    double t = t_init; 
    size_type step = 0; 

    d.dt = dt;


    // Set the external state variables
    setExternalStateVariable(d.s0, "Temperature", 293.15);
    setExternalStateVariable(d.s1, "Temperature", 293.15);

    update(d);

    // Set the imposed stresses
    const double SXX = 40.e6;
    const double SXY = 40.e6;

    const double ds = SXX / num_steps;

    d.s1.thermodynamic_forces[0] = 0.;
    d.s1.thermodynamic_forces[3] = 0.;

    // cout
    std::cout << t << ' ';
    cout_vector(d.s1.gradients);
    cout_vector(d.s1.thermodynamic_forces);
    cout_vector(d.s1.internal_state_variables);
    std::cout << std::endl;

    d.K[0] = 0; // random choice, I still don't understand how it works

    for (size_type i = 0; i < num_steps; ++i) {
        d.rdt = 1;

        // Update time and step
        t += dt;
        step++;

        // Perform integration for the current time step
        auto v = make_view(d);
        integrate(v, b);
        update(d);

        d.s1.thermodynamic_forces[0] += ds;
        d.s1.thermodynamic_forces[3] += ds;

        // cout
        std::cout << t << ' ';
        cout_vector(d.s1.gradients);
        cout_vector(d.s1.thermodynamic_forces);
        cout_vector(d.s1.internal_state_variables);
        std::cout << std::endl;

    }

    return 0;
}
```
The resulting output is the same for every time step after the first one :
(organized as : time, EXX, EYY, EZZ, EXY, EXZ, EYZ, SXX, SYY, SZZ, SXY, SXZ, SYZ, ElasticStrainXX, ElasticStrainYY, ElasticStrainZZ, ElasticStrainXY, ElasticStrainXZ, ElasticStrainYZ, p) 
```
0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
0.3 0 0 0 0 0 0 400000 0 0 400000 0 0 0 0 0 0 0 0 0 
```

Can someone provide an example of a working C++ code to do this ? 
Or maybe another example that could help me ?
In the long run, my goal would be to get stress values based on some imposed strain values. I would probably also need to compute the tangent matrix. Maybe, someone has an example adapted to this kind of use ?

Thanks in advance,

Lilou

**Comment by thelfer** [comment]
Hi @lilou-gautier,
Thanks for your interest in MFront and MGIS !
You won't be able reproduce MTest like this. MTest does some Newton iterations to find the appropriate strain at the end of the time step to satisfy the imposed stress.
I don't have any example at hand to do so (MGIS is meant to be integrated in an existing solver, not to become one !)
Best,
Thomas

**Comment by lilou-gautier** [comment]
Hello, 
First, thank you very much for your answer. 
You're right, I did not properly think about the steps behind MTest to compute the strain at each time step.
This was not an appropriate example of what I really need for my future use of MFront/MGIS.
Let me rephrase my question then.
Assuming I have the strain at one time step, I should be able to get the corresponding stress at this same time step using MFront/MGIS right ? This is what I struggle implementing in C++.
Do you have an example of this ? Or maybe indications on the steps to follow ?
Thanks again and sorry if this is still out of the scope of MGIS/MFront, I'm pretty lost.
Lilou

**Comment by thelfer** [comment]
The following example does that for one integration point: https://github.com/thelfer/MFrontGenericInterfaceSupport/blob/master/tests/IntegrateTest.cxx

This example does that for a set of integration points: https://github.com/thelfer/MFrontGenericInterfaceSupport/blob/master/tests/IntegrateTest2.cxx

**Comment by lilou-gautier** [comment]
Thank you. I had looked at these but I guess I mixed everything up when trying to reproduce the MTest example.
I should be able to find what I need.
Have a good week,
Lilou

**Comment by thelfer** [comment]
Do not hesitate if you need more help.
